File: pm-toolkit/scripts/proto-serve.js
#!/usr/bin/env node

/**
 * PM Toolkit -- Prototype Preview Server
 *
 * Serves Agent-generated HTML pages with a preview shell.
 *
 *   GET  /              -> preview shell (prototype.html)
 *   GET  /api/pages     -> list .html files in directory
 *   GET  /page/:name    -> serve raw HTML file (for iframe)
 *   DELETE /api/page/:name -> delete a page
 *
 * Usage: node proto-serve.js --dir <prototype-dir> [--port 9877]
 */

const http = require("http");
const fs = require("fs");
const path = require("path");

const args = process.argv.slice(2);
const getArg = (n, fb) => { const i = args.indexOf("--" + n); return i !== -1 && args[i + 1] ? args[i + 1] : fb; };

const DIR = path.resolve(getArg("dir", ""));
const PORT = parseInt(getArg("port", "9877"), 10);
const NAME_RE = /^[a-zA-Z0-9_-]+$/;

if (!DIR) { console.error("Usage: node proto-serve.js --dir <prototype-dir>"); process.exit(1); }
if (!fs.existsSync(DIR)) fs.mkdirSync(DIR, { recursive: true });

const SHELL = fs.readFileSync(path.join(__dirname, "prototype.html"), "utf-8");

function sendJson(res, code, data) {
    res.writeHead(code, { "Content-Type": "application/json; charset=utf-8", "Cache-Control": "no-store" });
    res.end(JSON.stringify(data));
}

function sendText(res, code, text) {
    res.writeHead(code, { "Content-Type": "text/plain; charset=utf-8", "Cache-Control": "no-store" });
    res.end(text);
}

function listPages() {
    return fs.readdirSync(DIR)
        .filter((f) => f.endsWith(".html"))
        .map((f) => {
            const name = f.replace(/\.html$/, "");
            const stat = fs.statSync(path.join(DIR, f));
            return { name, file: f, mtime: stat.mtimeMs };
        })
        .sort((a, b) => a.name.localeCompare(b.name));
}

const server = http.createServer((req, res) => {
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Access-Control-Allow-Methods", "GET,DELETE,OPTIONS");
    if (req.method === "OPTIONS") { res.writeHead(204); res.end(); return; }

    const url = new URL(req.url, `http://localhost:${PORT}`);

    // Preview shell
    if (url.pathname === "/" || url.pathname === "/index.html") {
        res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
        res.end(SHELL);
        return;
    }

    // List pages
    if (url.pathname === "/api/pages" && req.method === "GET") {
        sendJson(res, 200, { pages: listPages(), dir: DIR });
        return;
    }

    // Serve raw HTML page (for iframe)
    const pageMatch = url.pathname.match(/^\/page\/([a-zA-Z0-9_-]+)$/);
    if (pageMatch && req.method === "GET") {
        const file = path.join(DIR, pageMatch[1] + ".html");
        if (!fs.existsSync(file)) { sendText(res, 404, "Not found"); return; }
        res.writeHead(200, { "Content-Type": "text/html; charset=utf-8", "Cache-Control": "no-store" });
        res.end(fs.readFileSync(file, "utf-8"));
        return;
    }

    // Delete page
    const delMatch = url.pathname.match(/^\/api\/page\/([a-zA-Z0-9_-]+)$/);
    if (delMatch && req.method === "DELETE") {
        const name = delMatch[1];
        if (!NAME_RE.test(name)) { sendText(res, 400, "Invalid name"); return; }
        const file = path.join(DIR, name + ".html");
        try {
            if (fs.existsSync(file)) fs.unlinkSync(file);
            sendJson(res, 200, { ok: true });
        } catch (e) { sendText(res, 500, e.message); }
        return;
    }

    sendText(res, 404, "Not found");
});

server.listen(PORT, () => {
    const pages = listPages();
    console.log(`\nPrototype Preview`);
    console.log(`  Dir:   ${DIR}`);
    console.log(`  URL:   http://localhost:${PORT}`);
    console.log(`  Pages: ${pages.length}\n`);
});

process.on("SIGINT", () => { server.close(() => process.exit(0)); });
process.on("SIGTERM", () => { server.close(() => process.exit(0)); });

File: pm-toolkit/scripts/proto_serve.py
#!/usr/bin/env python3
"""PM Toolkit — Prototype Preview Server

Serves Agent-generated HTML pages with a preview shell.

  GET  /              → preview shell (prototype.html)
  GET  /api/pages     → list .html files in directory
  GET  /page/<name>   → serve raw HTML file (for iframe)
  DELETE /api/page/<name> → delete a page

Usage: python3 proto_serve.py --dir <prototype-dir> [--port 9877]
"""

import argparse
import json
import re
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path

NAME_RE = re.compile(r"^[a-zA-Z0-9_-]+$")
SCRIPT_DIR = Path(__file__).resolve().parent


def parse_args():
    p = argparse.ArgumentParser(description="PM Toolkit Prototype Preview Server")
    p.add_argument("--dir", required=True, help="Directory containing prototype HTML files")
    p.add_argument("--port", type=int, default=9877, help="Server port (default: 9877)")
    return p.parse_args()


def list_pages(proto_dir: Path) -> list[dict]:
    pages = []
    for f in sorted(proto_dir.iterdir()):
        if f.is_file() and f.suffix == ".html":
            name = f.stem
            mtime = f.stat().st_mtime * 1000
            pages.append({"name": name, "file": f.name, "mtime": mtime})
    return pages


class ProtoHandler(BaseHTTPRequestHandler):
    shell_html: str = ""
    proto_dir: Path = Path()

    def log_message(self, format, *args):
        pass

    def _cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET,DELETE,OPTIONS")
        self.send_header("Cache-Control", "no-store")

    def _send_json(self, code: int, payload: dict):
        body = json.dumps(payload, ensure_ascii=False).encode()
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self._cors_headers()
        self.end_headers()
        self.wfile.write(body)

    def _send_text(self, code: int, text: str, content_type: str = "text/plain; charset=utf-8"):
        body = text.encode()
        self.send_response(code)
        self.send_header("Content-Type", content_type)
        self._cors_headers()
        self.end_headers()
        self.wfile.write(body)

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors_headers()
        self.end_headers()

    def do_GET(self):
        path = self.path.split("?")[0]

        if path in ("/", "/index.html"):
            self._send_text(200, self.shell_html, "text/html; charset=utf-8")
            return

        if path == "/api/pages":
            self._send_json(200, {"pages": list_pages(self.proto_dir), "dir": str(self.proto_dir)})
            return

        m = re.match(r"^/page/([a-zA-Z0-9_-]+)$", path)
        if m:
            html_file = self.proto_dir / f"{m.group(1)}.html"
            if not html_file.exists():
                self._send_text(404, "Not found")
                return
            self._send_text(200, html_file.read_text(encoding="utf-8"), "text/html; charset=utf-8")
            return

        self._send_text(404, "Not found")

    def do_DELETE(self):
        path = self.path.split("?")[0]
        m = re.match(r"^/api/page/([a-zA-Z0-9_-]+)$", path)
        if m:
            name = m.group(1)
            if not NAME_RE.match(name):
                self._send_text(400, "Invalid name")
                return
            html_file = self.proto_dir / f"{name}.html"
            try:
                if html_file.exists():
                    html_file.unlink()
                self._send_json(200, {"ok": True})
            except Exception as e:
                self._send_text(500, str(e))
            return

        self._send_text(404, "Not found")


def main():
    args = parse_args()
    proto_dir = Path(args.dir).resolve()
    proto_dir.mkdir(parents=True, exist_ok=True)

    shell_path = SCRIPT_DIR / "prototype.html"
    shell_html = shell_path.read_text(encoding="utf-8") if shell_path.exists() else "<h1>prototype.html not found</h1>"

    ProtoHandler.shell_html = shell_html
    ProtoHandler.proto_dir = proto_dir

    server = HTTPServer(("", args.port), ProtoHandler)
    pages = list_pages(proto_dir)
    print(f"\nPrototype Preview")
    print(f"  Dir:   {proto_dir}")
    print(f"  URL:   http://localhost:{args.port}")
    print(f"  Pages: {len(pages)}\n")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        server.server_close()


if __name__ == "__main__":
    main()
